### DISCUSSION TITLE
Docker Quickstart - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#5072] [Docs]
# Docker Quickstart - Ultralytics YOLOv8 Docs

Complete guide to setting up and using Ultralytics YOLO models with Docker. Learn how to install Docker, manage GPU support, and run YOLO models in isolated containers.

https://docs.ultralytics.com/guides/docker-quickstart/

**Comment by glenn-jocher** [comment]
📣 New Docker Quickstart Guide is out! Got questions or need clarification? Ask away—We're here to help! 🐳

**Comment by MOXHAN** [comment]
Hi, I'm still new to Docker, so this might be a simple question. I ran the command, which was shown here, to pull the latest Dockerfile-jetson. Everything was successful and I could run the container. However the Dockerfile says the WORKDIR is usr/src/ultralytics, but I cannot find the ultralytics dir in the usr/src dir. 
1. Why is that? 
2. Can I change this to be in the directory I'm running my terminal in or should I just mount  a directory from there to the container?

Many thanks in advance!

**Reply by glenn-jocher** [reply]
Oh that's strange. My Dockerfile looks like this. When I run the Docker start command I enter directly into /usr/src/ultralytics:

```bash
root@6efe5504c7f8:/usr/src/ultralytics# ls
CITATION.cff     LICENSE      README.md        debug   docs      mkdocs.yml        runs       setup.py  ultralytics           yolov8n.pt
CONTRIBUTING.md  MANIFEST.in  README.zh-CN.md  docker  examples  requirements.txt  setup.cfg  tests     ultralytics.egg-info
```

Here are some usage examples run the Dockerfile and start in this directory, mounting a local volume as well with the -v argument:

```bash
# Build and Push
t=ultralytics/ultralytics:latest && sudo docker build -f docker/Dockerfile -t $t . && sudo docker push $t

# Pull and Run with access to all GPUs
t=ultralytics/ultralytics:latest && sudo docker pull $t && sudo docker run -it --ipc=host --gpus all $t

# Pull and Run with access to GPUs 2 and 3 (inside container CUDA devices will appear as 0 and 1)
t=ultralytics/ultralytics:latest && sudo docker pull $t && sudo docker run -it --ipc=host --gpus '"device=2,3"' $t

# Pull and Run with local directory access
t=ultralytics/ultralytics:latest && sudo docker pull $t && sudo docker run -it --ipc=host --gpus all -v "$(pwd)"/datasets:/usr/src/datasets $t
```

**Comment by balfos** [comment]
Hi, (am very new to docker, and much else), in starting to try your ultralytics/ultralytics:jetson, the Dockerfile has FROM nvcr.io/nvidia/l4t-pytorch:r35.2.1-pth2.0-py3 but incompatibility messages get thrown when I try your example code from python3 within the container "results = model.track(source=....) " e.g. "WARNING ⚠️ torchvision==0.14 is incompatible with torch==2.0."   Nvidia says the l4t does bundle those two versions together, so not sure what best resolution is?  Other errors occur that might be from the l4t,  but I don't know enough to know, e.g. WARNING ⚠️ Environment does not support cv2.imshow() or PIL Image.show() and ultralytics requirement ['lapx>=0.5.2'] not found, attempting AutoUpdate.  thanks for any guidance. Regards, B

**Reply by MOXHAN** [reply]
Im getting the same warning about the incompatibility, also when using the jetson's Hello AI World docker and then installing ultralytics manually to use Yolov5. But so far it seems to be just a warning, not a problem, because I didnt have any problems using Yolov5 with Pytorch and running inference on it.

**Reply by pderrenger** [reply]
@MOXHAN hello! It's great to see your interest in using Ultralytics YOLOv8 with Docker on NVIDIA Jetson devices. The warning messages you're encountering are quite common when dealing with containerized environments, especially when specific versions of libraries are bundled together.

For the warning about `torchvision` and `torch` incompatibility, it's often just a notification and may not necessarily impact your use case if you're not using specific functionalities that are affected by the version mismatch. However, it's always good to use compatible versions to avoid any potential issues. You might want to check if there's an updated version of the Docker image that resolves this, or you could build your own Docker image with compatible versions.

Regarding the `cv2.imshow()` and `PIL Image.show()` warning, these functions are used to display images in a window, which might not be supported in a headless Docker environment or when you're accessing the container remotely. If you need to visualize the results, you could save the images to a file and view them outside the container, or set up X11 forwarding or a virtual frame buffer if you really need to display images within the container.

The warning about `lapx>=0.5.2` not being found suggests that the Linear Assignment Problem (LAP) library, which is used for data association in tracking, is either not installed or not at the required version. The AutoUpdate attempt is the system trying to resolve this by automatically installing the missing dependency. If this doesn't succeed, you might need to install it manually within the container.

Here's a general approach to address these issues:

1. **Incompatibility Warning**: Check for updates to the Docker image or consider building your own with compatible versions of `torch` and `torchvision`.

2. **Display Functions**: For headless environments, save the output images to disk and view them outside the container. Alternatively, set up a virtual display inside the container using tools like `xvfb`.

3. **Missing Dependencies**: If AutoUpdate doesn't resolve the missing `lapx` library, try installing it manually using `pip` within the container:
   ```bash
   pip install lapx --upgrade
   ```

4. **General Troubleshooting**: Ensure that you're using the latest version of the Ultralytics YOLOv8 repository, as updates and bug fixes are regularly pushed. If you're building your own Docker image, make sure to include all necessary dependencies in your Dockerfile.

Remember, when working with Docker, it's often helpful to have a bit of trial and error to fine-tune your environment. If you continue to encounter issues, consider reaching out to the Ultralytics team through the GitHub repository's issues section for more targeted support.

**Comment by barkhaaa** [comment]
Hello,

I'm looking for guidance on how to train a YOLOv8 model using the Dockerfile. Although I've successfully built the Docker image, I'm unsure about the specific command I need to use for training the model within the Docker container. Your assistance in this matter would be greatly appreciated.

Thank you.

**Reply by MOXHAN** [reply]
You can find instructions on using the CLI for training here: https://docs.ultralytics.com/usage/cli/
Remember to run these commands in the running container. :)

**Reply by barkhaaa** [reply]
Thank you for the reply it works.

**Reply by barkhaaa** [reply]
I have another question: If we wish to train a YOLOv8 model using our custom dataset, how can we achieve this using the Dockerfile we've created?

**Reply by MOXHAN** [reply]
You can pass the location of your data to the command with the data=your_data.yaml flag. You can see all available arguments here: https://docs.ultralytics.com/usage/cfg/

I only found documentation which shows the data as a .yaml file, I am unsure if it works with directories of images. However, I personally prefer training in a jupyter notebook, so you can also still edit and validate your training data if needed. Just use `pip install jupyter` in the container and you're ready to go. The link I provided in my answer before should also have documentation for commands in Python. Greetings :)

**Reply by barkhaaa** [reply]
Thanks

**Reply by pderrenger** [reply]
@barkhaaa you're welcome! If you have any more questions or need further assistance, feel free to ask. Happy training with YOLOv8! 😊🚀

**Comment by cnydo** [comment]
Hi, 

I'm learning Docker. How can I start logging Yolov8 experiment (with wandb, comet_ml, etc) on Docker 

Thank you!

**Reply by pderrenger** [reply]
@cnydo hello! To start logging your YOLOv8 experiments with tools like Weights & Biases (wandb), Comet ML, or similar platforms within a Docker container, you'll need to follow these general steps:

1. **Install the Logging Library:**
   Ensure that the logging library (wandb, comet_ml, etc.) is installed in your Docker environment. You can do this by including the installation command in your Dockerfile or by running it directly in the container.

   For example, to install wandb, you might add the following line to your Dockerfile:
   ```Dockerfile
   RUN pip install wandb
   ```

2. **Set Up Your Logging Account:**
   Create an account on the platform you wish to use (if you haven't already) and obtain your API key or token.

3. **Configure the Logging Library:**
   Configure the logging library with your API key or token. This can often be done by setting an environment variable or by logging in through the command line.

   For wandb, you can log in using:
   ```bash
   wandb login your_api_key
   ```
   You can also set the `WANDB_API_KEY` environment variable in your Dockerfile or pass it when running the container:
   ```bash
   docker run -e WANDB_API_KEY=your_api_key ...
   ```

4. **Modify Your Training Script:**
   Update your YOLOv8 training script to initialize and use the logging library. For wandb, you would typically add:
   ```python
   import wandb
   wandb.init(project="your_project_name")
   ```
   Ensure that your training script logs the desired metrics to the platform.

5. **Run Your Container:**
   Start your Docker container, ensuring that any necessary environment variables are passed in and that your training script is correctly set up to log metrics.

   For example:
   ```bash
   docker run -it --ipc=host --gpus all -v /path/on/host:/path/in/container -e WANDB_API_KEY=your_api_key your_docker_image
   ```

6. **Start Training:**
   Run your YOLOv8 training command within the Docker container, and it should automatically log the metrics to your chosen platform.

Remember to check the specific documentation for the logging tool you're using for any additional configuration steps or best practices. Happy logging! 🚀

**Comment by SolowYolow** [comment]
Hi, I'm trying to train a model using CLI within WSL, I can do 2 or 3 epochs before I get this error:

RuntimeError: CUDA error: out of memory
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Most discussions online tell me to alter the batch size, how would I do this within CLI?

Thanks

**Reply by pderrenger** [reply]
@SolowYolow hello! To adjust the batch size in the CLI, you can use the `batch` argument followed by the desired batch size value. For example, if you want to set the batch size to 8, you would include `batch=8` in your command. Here's how you can modify your training command:

```bash
yolo train data=your_dataset.yaml model=your_model.yaml epochs=3 batch=8
```

Make sure to replace `your_dataset.yaml` and `your_model.yaml` with the actual names of your dataset and model configuration files. Reducing the batch size can help alleviate the CUDA out of memory error by using less GPU memory during training. If you still encounter memory issues, consider reducing the batch size further.

**Comment by hgospodar** [comment]
why dose the code run in docker will looply print debug output：
```
0: 384x640 (no detections), 15.3ms
```
why does it run like that? infactly, i'm using stream mode to inference a RTSP video stream, and the debug output will take over my useful debug. How can i close it?

**Reply by glenn-jocher** [reply]
You can try `YOLO_VERBOSE=False yolo predict` in CLI to silence YOLO outputs.

**Comment by ice6** [comment]
Can I keep my dataset in the local machine, and using the HUB's other features?

**Reply by glenn-jocher** [reply]
@ice6 absolutely! You can keep your dataset on your local machine and still utilize the features provided by Ultralytics YOLOv8. When running YOLOv8 in a Docker container, you can mount your local dataset directory to the container using the `-v` or `--volume` flag. This way, the dataset remains on your machine, but the container can access and use it as if it were inside the container.

Here's an example command to mount your local dataset directory into a Docker container:

```bash
sudo docker run -it --ipc=host --gpus all -v /path/to/your/local/dataset:/path/in/container ultralytics/ultralytics:latest
```

Replace `/path/to/your/local/dataset` with the path to your dataset on your local machine, and `/path/in/container` with the path where you want the dataset to be accessible within the Docker container.

This setup allows you to train, validate, and predict using YOLOv8 while keeping your data on your local machine. 🚀

**Comment by RahafRafdan** [comment]
Could you recommend which Python version will work great with YOLO v8?

**Reply by glenn-jocher** [reply]
@RahafRafdan you can use Python 3.8 up to Python 3.11 with YOLOv8. We support all active versions of Python :)
